app/channel.go

```diff
@@ -322,69 +322,74 @@ func (a *App) CreateChannel(channel *model.Channel, addMember bool) (*model.Chan
 
 func (a *App) GetOrCreateDirectChannel(userId, otherUserId string) (*model.Channel, *model.AppError) {
 	channel, nErr := a.Srv().Store.Channel().GetByName("", model.GetDMNameFromIds(userId, otherUserId), true)
-	if nErr != nil {
-		var nfErr *store.ErrNotFound
-		if errors.As(nErr, &nfErr) {
-			var err *model.AppError
-			channel, err = a.createDirectChannel(userId, otherUserId)
-			if err != nil {
-				if err.Id == store.ChannelExistsError {
-					return channel, nil
-				}
-				return nil, err
-			}
-
-			a.WaitForChannelMembership(channel.Id, userId)
+	if nErr == nil {
```

Comment:
This change looks like a simplification, but IMO is against the idiomatic way of writing it. The idiomatic way is "the happy path is at the lowest level of indentation", you are short circuiting there to return the happy path right away and handling the error after that. I think is more readable to handle the error cases, because is what I going to expect.

Reply:
hmm, so are you suggesting to amended to something like this:

```go
	channel, nErr := a.Srv().Store.Channel().GetByName("", model.GetDMNameFromIds(userId, otherUserId), true)
	var nfErr *store.ErrNotFound
	if !errors.As(nErr, &nfErr) {
		return nil, model.NewAppError("GetOrCreateDirectChannel", "web.incoming_webhook.channel.app_error", nil, nErr.Error(), http.StatusInternalServerError)
	} else if nErr == nil {
		return channel, nil
	}
```

Since in the case that the error is notFound we need to create the channel. This method has 2 happy paths technically.

Reply:
actual the above will not work when the `nErr` is nil, so I will revert back to short-circuiting

Reply:
No, I'm suggesting something like keeping the old "big if clause" wrapping all the "error handling" case.

For me here, the happy path is "I try to get the existing channel, and is there, so I don't have to do anything, just returning the channel" (that should be at the lowest level of indentation), then you have "If I'm not able to get the channel, how I handle that case?" that should be inside the if body.

But this is again my own interpretation of how it should be. I always expect the `if err != nil` after something an error, so... having that `if err == nil`, maybe is going to confuse me (and probably I directly don't read that, a go straight forward to the if body expecting the error handling).

Reply:
Have no strong pref here, let me see what I can do in order to make it more readable, since IMO  the "big if clause" I do not think provides higher readability, maybe if I use local functions we can clean this up a bit. Let me see how that looks and we can take it from there.

Reply:
@jespino how that looks https://github.com/mattermost/mattermost-server/pull/16819/commits/5085aa64be956bb1c1618a869e1fe4c6338b99ad ?